### DISCUSSION TITLE
[general] Improved reporting of branches

### DISCUSSION THREAD

**Discussion by multimeric** [ropensci/targets#1306] [General]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I have a lot of dynamic branches, and my workflow often involves debugging errors in particular branches but not others. When I get a failure in one branch, it says something like `▶ dispatched branch instance_path_44d946c111c4f3e1`, and then it fails. However I'm not sure how to work backwards to determine which values of each upstream dependency were used in that branch. This is particularly bad in cases where the task fails *outside of the command* ([e.g. here](https://github.com/wlandau/crew.cluster/issues/43)), because then I don't even get a traceback.

I wonder if I could request a `reporter` that prints out the values of each dependency for dynamic branches. So it says something like:
```
▶ dispatched branch instance_path_44d946c111c4f3e1 using x = 1 and y = 3
```

**Comment by mattwarkentin** [comment]
Hi @multimeric,

I think workspaces (`tar_workspaces()` and `tar_workspace()`) and `tar_branches()` provide the functionality you need. Setting the option as:
```r
tar_option_set(workspace_on_error = TRUE)
```

Then if a dynamic branch fails, you can identify the saved "workspaces" with `tar_workspaces()` and then select the dynamic branch name and load it into the global environment for debugging with `tar_workspace("name_of_branch_1234")`. This will load all of the globals (functions and objects) required to debug your code/pipeline.

Lastly, if you want just the names of the upstream dependencies for branches, you can use `tar_branches()`:
```r
tar_branches(target_name, pattern = <<enter pattern here>>)
```

Lame example but showcases things:
``` r
targets::tar_dir({
  targets::tar_script({
    library(targets)

    foo <- function(x) {
      if (x == 5) rlang::abort('Something went wrong')
      x
    }

    tar_option_set(
      workspace_on_error = TRUE,
      error = 'continue'
    )

    list(
      tar_target(
        a,
        1:10
      ),
      tar_target(
        b,
        a,
        pattern = map(a)
      ),
      tar_target(
        c,
        foo(b),
        pattern = map(b)
      )
    )
  }, ask = FALSE)
  
  targets::tar_make()

  print(targets::tar_workspaces())
  # Could load workspace with `tar_workspace("c_24be1142")`

  targets::tar_branches(c, map(b))
})
#> ▶ dispatched target a
#> ● completed target a [0.001 seconds]
#> ▶ dispatched branch b_0a91b2ed
#> ● completed branch b_0a91b2ed [0 seconds]
#> ▶ dispatched branch b_847c5ae2
#> ● completed branch b_847c5ae2 [0 seconds]
#> ▶ dispatched branch b_1619f1a0
#> ● completed branch b_1619f1a0 [0 seconds]
#> ▶ dispatched branch b_9401bbb6
#> ● completed branch b_9401bbb6 [0 seconds]
#> ▶ dispatched branch b_d24115c8
#> ● completed branch b_d24115c8 [0 seconds]
#> ▶ dispatched branch b_fb63d300
#> ● completed branch b_fb63d300 [0 seconds]
#> ▶ dispatched branch b_3138aba9
#> ● completed branch b_3138aba9 [0 seconds]
#> ▶ dispatched branch b_ad3eda3e
#> ● completed branch b_ad3eda3e [0 seconds]
#> ▶ dispatched branch b_bf196919
#> ● completed branch b_bf196919 [0 seconds]
#> ▶ dispatched branch b_b0795e65
#> ● completed branch b_b0795e65 [0 seconds]
#> ● completed pattern b
#> ▶ dispatched branch c_66acaf33
#> ● completed branch c_66acaf33 [0 seconds]
#> ▶ dispatched branch c_02f9cbce
#> ● completed branch c_02f9cbce [0.001 seconds]
#> ▶ dispatched branch c_c46c3a46
#> ● completed branch c_c46c3a46 [0 seconds]
#> ▶ dispatched branch c_0b5d87dc
#> ● completed branch c_0b5d87dc [0 seconds]
#> ▶ dispatched branch c_24be1142
#> ▶ recorded workspace c_24be1142
#> ✖ errored branch c_24be1142
#> Error: Something went wrong
#> 
#> ▶ dispatched branch c_0ff7342d
#> ● completed branch c_0ff7342d [0 seconds]
#> ▶ dispatched branch c_7de5013b
#> ● completed branch c_7de5013b [0 seconds]
#> ▶ dispatched branch c_54814e50
#> ● completed branch c_54814e50 [0 seconds]
#> ▶ dispatched branch c_241b47cd
#> ● completed branch c_241b47cd [0.001 seconds]
#> ▶ dispatched branch c_419518f2
#> ● completed branch c_419518f2 [0 seconds]
#> ✖ errored pipeline [0.167 seconds]
#> [1] "c_24be1142"
#> # A tibble: 10 × 2
#>    c          b         
#>    <chr>      <chr>     
#>  1 c_66acaf33 b_0a91b2ed
#>  2 c_02f9cbce b_847c5ae2
#>  3 c_c46c3a46 b_1619f1a0
#>  4 c_0b5d87dc b_9401bbb6
#>  5 c_24be1142 b_d24115c8
#>  6 c_0ff7342d b_fb63d300
#>  7 c_7de5013b b_3138aba9
#>  8 c_54814e50 b_ad3eda3e
#>  9 c_241b47cd b_bf196919
#> 10 c_419518f2 b_b0795e65
```

**Reply by mattwarkentin** [reply]
That being said, @wlandau, why does the `pattern` argument need to be supplied to `tar_branches()`. Couldn't this be inferred from the list of targets using the target `name`? It _feels_ like I should be able to just call `tar_branches(c)`.

**Reply by wlandau** [reply]
To get that list of targets, we would need to run `_targets.R`, which might take extra time and run code that the user is not expecting to run for such a lightweight function.

**Reply by multimeric** [reply]
Thanks for the answers.

`workspace_on_error = TRUE` is a nice suggestion, and probably a good general solution to this problem. However it will presumably save the entire environment which in my case is very large: hundreds of GB in my case. You could argue that a reporter would probably also have issues in this scenario which is probably fair.

I've looked into `tar_branches()` before. In my current pipeline I get `! number of predicted branches (26) disagrees with number of past observed branches (18)` which means I get no output. In the past I've wanted it to read `_targets.R` as discussed above, especially because with static branching, I don't even know what the failing target is called because its name is generated by `tar_map`. So I'd love a `tar_branches(read_pipeline = TRUE)` that takes a performance hit but is easier to use.

Now that I've thought about this more, I wonder if I could just make a target factory that adds a configurable logging before the main task.

**Reply by wlandau** [reply]
> However it will presumably save the entire environment which in my case is very large:

It only saves lightweight references to the objects that need to be loaded. The workspace file itself is extremely light, regardless of what is in the actual data.

**Reply by multimeric** [reply]
Thanks, I'll give this a try then. But if targets has a concept of "references", then this might have a lot of benefits elsewhere. When I teach `targets` I get a lot of questions about how to avoid copying large objects, for example when you want to take several targets, put them into a list and use that as a new target. Does this concept help with that problem?

**Reply by mattwarkentin** [reply]
> So I'd love a tar_branches(read_pipeline = TRUE) that takes a performance hit but is easier to use.

I would like something this too. Whether a new function or a configuration of the current function. Is this something you would consider @wlandau?

**Reply by wlandau** [reply]
Implemented in a131acf2a31f8767185544930d13ef46720b20fc.

**Reply by multimeric** [reply]
Just noting that it doesn't currently seem to handle nested targets. I tested the latest commit #f151c64a48, and I can reproduce the issue with the following `_targets.R`:

```R
library(targets)
list(
  tar_target(foo, 1),
  list(
    tar_target(bar, 2)
  )
)
```
```
> tar_branches()
Error:
! Found a non-target object in the target list. The target script file (default: _targets.R) must end with a list of tar_target() objects (recommended) or a tar_pipeline() object (deprecated).
```

**Reply by wlandau** [reply]
Should be fixed now.

**Reply by multimeric** [reply]
Unfortunately I now get the following error with the above example script:
```
Error:
!  is not a target in _targets.R
```

**Reply by wlandau** [reply]
Should fix itself if you supply a `name` argument.

**Reply by multimeric** [reply]
Thanks. However, because the target that I want to query hasn't completed all branches, it uses the incomplete metadata, and I get this:

```R
> tar_branches(name = instance_path)
Error in data.frame(..., check.names = FALSE) : 
  arguments imply differing number of rows: 18, 26
```

This is because `children` (which is derived from metadata, and so is missing some rows), is shorter than `out` (which is derived form the pipeline spec only)

**Reply by wlandau** [reply]
Unfortunately we can't get around this limitation (and others) of `tar_branches()`. `tar_branches()` is only an approximation.
